### ISSUE TITLE
Milkv duo support NAND image

### ISSUE THREAD

**Issue by Markus250** [riscv/meta-riscv#487]
In buildroot I can compile image for SPI NAND next command:
./build.sh milkv-duo-spinand
and in the folder out/
I will receive the files for the firmware nand (boot.spinand, fip.bin, rootfs.spinand, system.spinand)
I load this into nand and it works
How can create an image for nand using this layer (yocto)?

sorry for my eng

**Comment by Markus250** [comment]
Thanks

**Comment by ziswiler** [comment]
This broke the build for me:

```
make: Leaving directory '/var/home/sumo/Sources/riscv-yocto/build/tmp/work/riscv64-poky-linux/milkv-duo-fsbl/1.0/git/firmware'
usage: fiptool.py [-h] [-v] {genfip} ...
fiptool.py: error: unrecognized arguments: --NAND_BOOT=1
WARNING: /var/home/sumo/Sources/riscv-yocto/build/tmp/work/riscv64-poky-linux/milkv-duo-fsbl/1.0/temp/run.do_compile.1911581:150 exit 2 from 'python3 /var/home/sumo/Sources/riscv-yocto/build/tmp/work/riscv64-poky-linux/milkv-duo-fsbl/1.0/git/firmware/plat/cv180x/fiptool.py genfip /var/home/sumo/Sources/riscv-yocto/build/tmp/work/riscv64-poky-linux/milkv-duo-fsbl/1.0/git/firmware/build/fip.bin --MONITOR_RUNADDR=0x80000000 --CHIP_CONF=/var/home/sumo/Sources/riscv-yocto/build/tmp/work/riscv64-poky-linux/milkv-duo-fsbl/1.0/git/firmware/plat/cv180x/chip_conf.bin --NOR_INFO=FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF --NAND_INFO=00000000 --BL2=/var/home/sumo/Sources/riscv-yocto/build/tmp/work/riscv64-poky-linux/milkv-duo-fsbl/1.0/git/firmware/build/cv180x/bl2.bin --BLCP_IMG_RUNADDR=0x05200200 --BLCP_PARAM_LOADADDR=0 --BLCP_2ND=/var/home/sumo/Sources/riscv-yocto/build/tmp/work/riscv64-poky-linux/milkv-duo-fsbl/1.0/git/firmware/build/blank.bin --BLCP_2ND_RUNADDR=0x83f40000 --DDR_PARAM=/var/home/sumo/Sources/riscv-yocto/build/tmp/work/riscv64-poky-linux/milkv-duo-fsbl/1.0/git/firmware/test/cv181x/ddr_param.bin --MONITOR=/var/home/sumo/Sources/riscv-yocto/build/tmp/deploy/images/milkv-duo/fw_dynamic.bin --LOADER_2ND=/var/home/sumo/Sources/riscv-yocto/build/tmp/deploy/images/milkv-duo/u-boot.bin --NAND_BOOT=1'
WARNING: Backtrace (BB generated script): 
        #1: do_compile, /var/home/sumo/Sources/riscv-yocto/build/tmp/work/riscv64-poky-linux/milkv-duo-fsbl/1.0/temp/run.do_compile.1911581, line 150
        #2: main, /var/home/sumo/Sources/riscv-yocto/build/tmp/work/riscv64-poky-linux/milkv-duo-fsbl/1.0/temp/run.do_compile.1911581, line 202

```

**Comment by Markus250** [comment]
Yep, It doesn't work, I have exactly the same error
I am check fiptool.py args, he does not have --NAND_BOOT=1
@vishwamartur you make patch forfiptool.py?

**Comment by kraj** [comment]
@ziswiler @Markus250 I have removed the non existing option for now. I assume there is a patch for https://github.com/milkv-duo/duo-buildroot-sdk/blob/develop/fsbl/plat/cv181x/fiptool.py which is missing, or the option was added wrongly.

**Comment by Markus250** [comment]
@kraj  Yes, without this argument, the assembly of the image has advanced. Also, I had to comment the line 
`cp ${WORKDIR}/fip.bin ${DEPLOYDIR}/` in file `milkv-duo-bootfiles.bb` because in recipe `milkv-duo-bootfiles.bb` and `milkv-duo-fsbl.bb` simultaneous copying of the file `fip.bin`
But it still did not work, files `rootfs.spinand`, `boot.spinand`, etc..  do not appear, do not even gather =(
How much I could make out buildroot this files make python script mkubiimg.py, example from `duo-buildroot-sdk/build/Makefile`:
`${Q}python3 $(COMMON_TOOLS_PATH)/spinand_tool/mkubiimg.py --ubionly $(FLASH_PARTITION_XML) ROOTFS $(OUTPUT_DIR)/rawimages/rootfs.sqsh $(OUTPUT_DIR)/rawimages/rootfs.spinand -b $(CONFIG_NANDFLASH_BLOCKSIZE) -p $(CONFIG_NANDFLASH_PAGESIZE)` 
Again, these are all my assumptions, I could never collect the image of NAND through Yocto 
I will be grateful if you help me)

**Comment by kraj** [comment]
@Markus250 does the fix proposed in #511 fix the issue ?

**Comment by Markus250** [comment]
@kraj  Well, the assembly is going on, but the image for nand not created =(

**Comment by kraj** [comment]
> [@kraj](https://github.com/kraj) Well, the assembly is going on, but the image for nand not created =(

That seems not good. Does it work if we revert the original patch to add NAND support ?

**Comment by Markus250** [comment]
> @kraj That seems not good. Does it work if we revert the original patch to add NAND support ?

Unfortunately, no, if you look at how the buildroot does it, then there is more manipulations there, I tried to do it myself, but nothing worked for me

**Comment by kraj** [comment]
@Markus250 ok, are you able to build if you revert https://github.com/riscv/meta-riscv/commit/b94502a53caa8efe101a1567eadea8bccdc60198 ?

in meta-riscv you can do

git revert b94502a53caa8efe101a1567eadea8bccdc60198

and build

**Comment by Markus250** [comment]
@kraj Yes, but need we must change the following:
`conf/machine/milkv-duo.conf` restore `WKS_FILE ?= "milkv-duo.wks"` instead of `WKS_FILE ?= "milkv-duo-spinand.wks"`
revert old milkv-duo-fsbl.bb from meta-riskv branch scarthgap
it seems like everything, then you can collect an image, but again, this image is not for spinand
in branch scarthgap image build successfully for sd-card, It is very sorry that no one can add the opportunity to collect under spi-nand =(

**Comment by kraj** [comment]
if it is breaking then I would like to revert it. Can you post the patch that works for you ?
